kernel/src/log_segment.rs

```diff
@@ -112,13 +124,235 @@ impl LogSegment {
+    pub(crate) fn list_log_files_from_version(
+        fs_client: &dyn FileSystemClient,
+        log_root: &Url,
+        version: Option<Version>,
+    ) -> DeltaResult<(Vec<ParsedLogPath>, Vec<ParsedLogPath>, i64)> {
+        let begin_version = version.unwrap_or(0);
+        let version_prefix = format!("{:020}", begin_version);
+        let start_from = log_root.join(&version_prefix)?;
+
+        let mut max_checkpoint_version = version.map_or(-1, |x| x as i64);
+        let mut checkpoint_files = vec![];
+        // We expect 10 commit files per checkpoint, so start with that size. We could adjust this based
+        // on config at some point
+        let mut commit_files = Vec::with_capacity(10);
+
+        for meta_res in fs_client.list_from(&start_from)? {
+            let meta = meta_res?;
+            let parsed_path = ParsedLogPath::try_from(meta)?;
+            // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
+            if let Some(parsed_path) = parsed_path {
+                if parsed_path.is_commit() {
+                    commit_files.push(parsed_path);
+                } else if parsed_path.is_checkpoint() {
+                    let path_version = parsed_path.version as i64;
+                    match path_version.cmp(&max_checkpoint_version) {
+                        Ordering::Greater => {
+                            max_checkpoint_version = path_version;
+                            checkpoint_files.clear();
+                            checkpoint_files.push(parsed_path);
+                        }
+                        Ordering::Equal => checkpoint_files.push(parsed_path),
+                        Ordering::Less => {}
+                    }
+                }
+            }
+        }
+
+        debug_assert!(
+            commit_files
+                .windows(2)
+                .all(|cfs| cfs[0].version <= cfs[1].version),
+            "fs_client.list_from() didn't return a sorted listing! {:?}",
+            commit_files
+        );
+
+        Ok((commit_files, checkpoint_files, max_checkpoint_version))
+    }
+
+    /// List all log files after a given checkpoint.
+    pub(crate) fn list_log_files_with_checkpoint(
+        checkpoint_metadata: &CheckpointMetadata,
+        fs_client: &dyn FileSystemClient,
+        log_root: &Url,
+    ) -> DeltaResult<(Vec<ParsedLogPath>, Vec<ParsedLogPath>)> {
+        let (commit_files, checkpoint_files, max_checkpoint_version) =
+            Self::list_log_files_from_version(
+                fs_client,
+                log_root,
+                Some(checkpoint_metadata.version),
+            )?;
+
+        if checkpoint_files.is_empty() {
+            // TODO: We could potentially recover here
+            return Err(Error::generic(
+                "Had a _last_checkpoint hint but didn't find any checkpoints",
+            ));
+        }
+
+        if max_checkpoint_version != checkpoint_metadata.version as i64 {
```

Comment:
```suggestion
        if max_checkpoint_version != checkpoint_metadata.version.try_into()? {
```

Reply:
can we just make max_checkpoint_version a `Version`? or option version? which is u64